File: leet_LIST_0876_mid_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head.next:
            return head
        if not head.next.next:
            return head.next
        f, s = head, head
        while f.next and f.next.next:
            s = s.next
            f = f.next.next
        if f.next:
            return s.next
        return s

File: leet_LIST_0876_mid_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {

    Solutions := []func(*ListNode) *ListNode {
        // Naive,
        Fast_and_slow_pointers,
    }
    return Solutions[0](head)
}

func Fast_and_slow_pointers(head *ListNode) *ListNode {

    f,s := head, head
    for f != nil && f.Next != nil {
        f = f.Next.Next
        s = s.Next
    }
    if f != nil && f.Next != nil {
        s = s.Next
    }
    return s
}

func Naive(head *ListNode) *ListNode {
    size := 0
    node := head
    for node != nil {
        node = node.Next
        size++
    }
    node = head
    size /= 2
    for size > 0 {
        node = node.Next
        size--
    }
    return node
}
